pkg/datastore/credentials.go

```diff
@@ -0,0 +1,87 @@
+package datastore
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awsconfig "github.com/aws/aws-sdk-go-v2/config"
+	rdsauth "github.com/aws/aws-sdk-go-v2/feature/rds/auth"
+	"golang.org/x/exp/maps"
+
+	log "github.com/authzed/spicedb/internal/logging"
+)
+
+// CredentialsProvider allows datastore credentials to be retrieved dynamically
+type CredentialsProvider interface {
+	// Name returns the name of the provider
+	Name() string
+	// Get returns the username and password to use when connecting to the underlying datastore
+	Get(ctx context.Context, dbHostname string, dbPort uint16, dbUser string) (string, string, error)
+}
+
+var NoCredentialsProvider CredentialsProvider = nil
+
+type credentialsProviderBuilderFunc func(ctx context.Context) (CredentialsProvider, error)
+
+const (
+	// AWSIAMCredentialProvider generates AWS IAM tokens for authenticating with the datastore (i.e. RDS)
+	AWSIAMCredentialProvider = "aws-iam"
+)
+
+var BuilderForCredentialProvider = map[string]credentialsProviderBuilderFunc{
+	AWSIAMCredentialProvider: newAWSIAMCredentialsProvider,
+}
+
+// CredentialsProviderOptions returns the full set of credential provider names, sorted and quoted into a string.
+func CredentialsProviderOptions() string {
+	ids := maps.Keys(BuilderForCredentialProvider)
+	sort.Strings(ids)
+	quoted := make([]string, 0, len(ids))
+	for _, id := range ids {
+		quoted = append(quoted, `"`+id+`"`)
+	}
+	return strings.Join(quoted, ", ")
+}
+
+// NewCredentialsProvider create a new CredentialsProvider for the given name
+// returns an error if no match is found, of if there is a problem creating the given CredentialsProvider
+func NewCredentialsProvider(ctx context.Context, name string) (CredentialsProvider, error) {
+	if name == "" {
```

Comment:
the default value for the cmd flag is `""`, so we treat this as a valid option and return the `NoCredentialsProvider`

I can add a comment to this effect, or take a different approach?

Reply:
Ideally, the empty string is handled at the caller and NewCredentialsProvider is never called with an invalid name

Reply:
sounds good, will go with that then